apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1217 @@
+            const CSMWorld::LandNormalsColumn::DataType landNormalsPointer = landTable.data(landTable.getModelIndex(cellId, landnormalsColumn)).value<CSMWorld::LandNormalsColumn::DataType>();
+            CSMWorld::LandNormalsColumn::DataType landNormalsNew(landNormalsPointer);
+
+            for(int i = 0; i < ESM::Land::LAND_SIZE; ++i)
+            {
+                for(int j = 0; j < ESM::Land::LAND_SIZE; ++j)
+                {
+                    float v1[3];
+                    float v2[3];
+                    float normal[3];
+                    float hyp;
+
+                    v1[0] = 128;
+                    v1[1] = 0;
+                    if (i < ESM::Land::LAND_SIZE - 1) v1[2] = landShapePointer[j * ESM::Land::LAND_SIZE + i + 1] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                    else
+                    {
+                        bool noCell = document.getData().getCells().searchId (CSMWorld::CellCoordinates::generateId(cellCoordinates.getX() + 1, cellCoordinates.getY())) == -1;
+                        bool noLand = document.getData().getLand().searchId (CSMWorld::CellCoordinates::generateId(cellCoordinates.getX() + 1, cellCoordinates.getY())) == -1;
+                        if (!noLand && !noCell)
+                            v1[2] = landRightShapePointer[j * ESM::Land::LAND_SIZE + 1] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                        else
+                            v1[2] = 0;
+                    }
+
+                    v2[0] = 0;
+                    v2[1] = 128;
+                    if (j < ESM::Land::LAND_SIZE - 1) v2[2] = landShapePointer[(j + 1) * ESM::Land::LAND_SIZE + i] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                    else
+                    {
+                        bool noCell = document.getData().getCells().searchId (CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY() + 1)) == -1;
+                        bool noLand = document.getData().getLand().searchId (CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY() + 1)) == -1;
+                        if (!noLand && !noCell)
+                            v2[2] = landDownShapePointer[ESM::Land::LAND_SIZE + i] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                        else
+                            v2[2] = 0;
+                    }
+
+                    normal[1] = v1[2]*v2[0] - v1[0]*v2[2];
+                    normal[0] = v1[1]*v2[2] - v1[2]*v2[1];
+                    normal[2] = v1[0]*v2[1] - v1[1]*v2[0];
+
+                    hyp = sqrt(normal[0]*normal[0] + normal[1]*normal[1] + normal[2]*normal[2]) / 127.0f;
+
+                    normal[0] /= hyp;
+                    normal[1] /= hyp;
+                    normal[2] /= hyp;
```

Comment:
Consider to use `osg::Vec3f` for such kind of calculations.

Reply:
I have no objections against it, but I do not understand the reason. Vec3f is container for three floats, right? Are there special advantages for using it?

Reply:
It's more readable as it's saying you're working on a vector instead of just three floats. It's also less typing as you have operators that work on a vector instead of just its components.

Reply:
Committed in ce061d9 . I was not able to change the cross product to use osg * operator. There's something with value_type float, e.g. `normal = v1 * v2 ` doesn't compile.

Reply:
Cross product is `^`, isn't it? `*` is dot product, which returns a scalar, so of course that'll give you errors.